addons/stock/models/stock_move.py

```diff
@@ -1816,6 +1821,26 @@ def _create_extra_move(self):
                 extra_move = extra_move._action_confirm()
         return extra_move | self
 
+    def _check_unlink_move_dest(self):
+        """ For each move in self, check if the location_dest_id of move is outside
+            (!= and not a child of) the source location of move_dest_ids,
+            if so, break the link.
+        """
+        new_push_moves = self.env['stock.move']
+        moves_to_mts = self.env['stock.move']
+        for move in self:
+            move_dest_ids_to_unlink = move.move_dest_ids.filtered(
+                lambda m: move.location_dest_id != m.location_id and str(m.location_id.id) not in move.location_dest_id.parent_path.split('/')
+            )
+            move_dest_ids_to_unlink.move_orig_ids = [Command.unlink(move.id)]
+            moves_to_mts |= move_dest_ids_to_unlink
+            if moves_to_mts:
+                new_push_moves |= move._push_apply()
```

Comment:
You can not move it outside the loop to batch the push_apply?

Reply:
We should only run the `push_apply` if that move has dest_ids unlinked, if we have more than one move in self and only one move was unlinked from its dest move, both moves will have push rules applied.
I should actually check for `move_dest_ids_to_unlink` instead :thinking:

Reply:
I mean that we can group them outside the initial loop. If move has move_dest_ids_to_unlink then we could remember it and launch _push_apply only once. It could be great for the perf since we will not search too much for the push rules